components/net/subresource_integrity.rs

```diff
@@ -0,0 +1,164 @@
+        SriEntry {
+            alg: alg.to_owned(),
+            val: val.to_owned(),
+            opt: opt,
+        }
+    }
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#parse-metadata
+pub fn parsed_metadata(integrity_metadata: &str) -> Vec<SriEntry> {
+    // Step 1
+    let mut result = vec![];
+
+    // Step 3
+    let tokens: Vec<&str> = integrity_metadata.split(" ").collect();
+
+    for token in tokens {
+        let parsed_data: Vec<&str> = token.split("-").collect();
+
+        if parsed_data.len() > 0 {
+            let alg = parsed_data[0];
+
+            if SUPPORTED_ALGORITHM.contains_key(alg) {
+                let data: Vec<&str> = parsed_data[1].split("?").collect();
+                let digest = data[0];
+
+                let opt = if data.len() > 1 {
+                    Some(data[1].to_owned())
+                } else {
+                    None
+                };
+
+                result.push(SriEntry::new(alg, digest, opt))
+            }
+        }
+    }
+
+    return result;
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#getprioritizedhashfunction
+pub fn get_prioritized_hash_function(hash_func_left: &str, hash_func_right: &str) -> String {
+    let left_priority = SUPPORTED_ALGORITHM.get(hash_func_left).unwrap();
+    let right_priority = SUPPORTED_ALGORITHM.get(hash_func_right).unwrap();
+
+    if left_priority == right_priority {
+        return "".to_owned();
+    }
+    if left_priority > right_priority {
+        hash_func_left.to_owned()
+    } else {
+        hash_func_right.to_owned()
+    }
+
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#get-the-strongest-metadata
+pub fn get_strongest_metadata(integrity_metadata_list: Vec<SriEntry>) -> Vec<SriEntry> {
+    let mut result: Vec<SriEntry> = vec![integrity_metadata_list[0].clone()];
+    let mut current_algorithm = result[0].alg.clone();
+
+    for i in 1..integrity_metadata_list.len() {
+        let intgerity_metadata = integrity_metadata_list[i].clone();
+
+        let prioritized_hash = get_prioritized_hash_function(&intgerity_metadata.alg,
+                                                                 &*current_algorithm);
+        if prioritized_hash == "" {
+            result.push(intgerity_metadata.clone());
+        } else if prioritized_hash != *current_algorithm {
+            result = vec![intgerity_metadata.clone()];
+            current_algorithm = prioritized_hash;
+        }
+    }
+
+    result
+}
+
+/// https://w3c.github.io/webappsec-subresource-integrity/#apply-algorithm-to-response
+fn apply_algorithm_to_response(body: MutexGuard<ResponseBody>,
+                               message_digest: MessageDigest)
+                               -> String {
+    if let ResponseBody::Done(ref vec) = *body {
+        let response_digest = hash(message_digest, vec);
+        response_digest.to_base64(STANDARD)
+    } else {
+        "".to_owned()
```

Comment:
We should probably make this `unreachable!("Tried to calculate digest of incomplete response body")` instead.